dll/win32/shell32/propsheet.cpp

```diff
@@ -0,0 +1,178 @@
+        if (hr > 0) // The returned index is one-based (relative to this extension).
+        {
+            hr += OldCount;
+            psh.nStartPage = hr - 1;
+        }
+    }
+    return hr;
+}
+
+static HRESULT
+SHELL_CreatePropSheetStubWindow(CStubWindow32 &stub, PCIDLIST_ABSOLUTE pidl, const POINT *pPt)
+{
+    PWSTR Path;
+    if (!pidl || FAILED(SHGetNameFromIDList(pidl, SIGDN_DESKTOPABSOLUTEPARSING, &Path)))
+        Path = NULL; // If we can't get a path, we simply will not be able to reuse this window
+
+    HRESULT hr = CStubWindow32::CreateStub(stub, CStubWindow32::TYPE_PROPERTYSHEET, Path, pPt);
+    SHFree(Path);
+    UINT flags = SHGFI_ICON | SHGFI_ADDOVERLAYS;
+    SHFILEINFO sfi;
+    if (SUCCEEDED(hr) && SHGetFileInfoW((LPWSTR)pidl, 0, &sfi, sizeof(sfi), SHGFI_PIDL | flags))
+        stub.SendMessage(WM_SETICON, ICON_BIG, (LPARAM)sfi.hIcon);
+    return hr;
+}
+
+/*************************************************************************
+ *  SHELL32_PropertySheet [INTERNAL]
+ *  PropertySheetW with stub window.
+ */
+static INT_PTR
+SHELL32_PropertySheet(LPPROPSHEETHEADERW ppsh, IDataObject *pDO)
+{
+    CStubWindow32 stub;
+    POINT pt, *ppt = NULL;
+    if (pDO && SUCCEEDED(DataObject_GetOffset(pDO, &pt)))
+        ppt = &pt;
+    PIDLIST_ABSOLUTE pidl = SHELL_DataObject_ILCloneFullItem(pDO, 0);
+    HRESULT hr = SHELL_CreatePropSheetStubWindow(stub, pidl, ppt);
+    ILFree(pidl);
+    if (FAILED(hr))
+        return hr == HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS) ? 0 : -1;
+
+    INT_PTR Result = -1;
+    ppsh->hwndParent = stub;
+    if (ppsh->nPages)
+    {
+        Result = PropertySheetW(ppsh);
+    }
+    else
+    {
+        WCHAR szFormat[128], szMessage[_countof(szFormat) + 42];
+        LoadStringW(shell32_hInstance, IDS_CANTSHOWPROPERTIES, szFormat, _countof(szFormat));
+        wsprintfW(szMessage, szFormat, ERROR_CAN_NOT_COMPLETE);
+        MessageBoxW(NULL, szMessage, NULL, MB_ICONERROR);
+    }
+    stub.DestroyWindow();
+    return Result;
+}
+
+/*************************************************************************
+ *  SHELL32_OpenPropSheet [INTERNAL]
+ *  The real implementation of SHOpenPropSheetW.
+ */
+static BOOL
+SHELL32_OpenPropSheet(LPCWSTR pszCaption, HKEY *ahKeys, UINT cKeys,
+                      const CLSID *pclsidDefault, IDataObject *pDO, LPCWSTR pStartPage)
+{
+    HKEY hKeyProgID = cKeys ? ahKeys[cKeys - 1] : NULL; // Windows uses the last key for some reason
+    HPROPSHEETPAGE hppages[MAX_PROPERTY_SHEET_PAGE];
+    PROPSHEETHEADERW psh = { sizeof(psh), PSH_PROPTITLE };
+    psh.phpage = hppages;
+    psh.hInstance = shell32_hInstance;
+    psh.pszCaption = pszCaption;
+    psh.pStartPage = pStartPage;
+
+    if (pclsidDefault)
+        AddPropSheetHandlerPages(*pclsidDefault, pDO, hKeyProgID, psh);
+
+    for (UINT i = 0; i < cKeys; ++i)
+    {
+        // Note: We can't use SHCreatePropSheetExtArrayEx because we need the AddPages return value.
```

Comment:
Where are you retrieving this AddPages return value here? Do you mean, the value returned by the SHELL32_PropertySheet() call below?

Reply:
There is a public function called `SHCreatePropSheetExtArrayEx` but it cannot be used here because it does not give us the return value of `IShellPropSheetExt::AddPages` which we need so we can set the correct initial page (open the properties of a .lnk file to see this in action).

I will update the comment to try to make it clear which return value I'm talking about.